[DOC-13536] USING AI - UI support and other improvements (#417)
couchbaselabs/docs-devex: modules/n1ql/pages/n1ql-language-reference/using-ai.adoc

Reviewer: Almost forgot, do you want to add a **See Also** section linking to the [Couchbase Capella operational docs](https://docs.couchbase.com/cloud/get-started/intro.html)? Since they need to have a Capella account to use this feature.

Author: Yes, added it now. Thanks.

== Purpose

You can use the USING AI statement to convert a natural language prompt into a {sqlpp} query.
When executed, the statement passes the input to Large Language Models (LLMs), which interpret the request and return the equivalent {sqlpp} query. 

For example, you can input prompts such as `How many airlines are based in Europe` or `List the names of all hotels in the same city as an airport`, and the statement generates the corresponding SQL++ query. 

If the generated statement is a SELECT query, the Query Service automatically executes it and returns the results.
For all other query types, it returns the generated statement as a string without executing it. 
However, you can modify this behavior by using the <<optional-parameters, execute>> option.

IMPORTANT: With the introduction of the USING AI statement, you might notice that AI is now recognized as a keyword.
However, this change has been implemented in a way that does not break existing queries or applications where "ai" is used as a field name or identifier.
For example, queries like `SELECT ai FROM XYZ` will continue to work without any issues.

> AI is treated as a keyword only when used in conjunction with the USING AI statement.

Reviewer: I would just dial back the "marketing" tone a little — all the talk about changes being new and exciting go out of date as soon as the next version is released

```suggestion
IMPORTANT: The word `AI` is recognized as a keyword, but only when used as part of the `USING AI` statement.
When used by itself as a field name or identifier, you do not need to escape the word `AI` by enclosing it in backticks.
For example, in a query like `SELECT ai FROM XYZ`, you can use `ai` as a field name without needing to escape it.
```

> [Required] A natural language request that you want to convert into a SQL++ query.

Reviewer: ```suggestion
[Required] A natural language request that you want to convert into a {sqlpp} query.
```

options::
[Optional] A JSON object specifying additional <<optional-parameters,options>> for the statement.
+

> By default, the statement uses the xref:n1ql-rest-query:index.adoc#Request[natural_orgid], xref:n1ql-rest-query:index.adoc#Request[natural_cred], and xref:n1ql-rest-query:index.adoc#Request[natural_context] parameters.

Reviewer: Just reinforcing that these are request-level parameters, they're not part of this statement, which is why they're not described in this page. Also you can link directly to the options!

```suggestion
By default, the statement uses the xref:n1ql-rest-query:index.adoc#natural_orgid[natural_orgid], xref:n1ql-rest-query:index.adoc#natural_cred[natural_cred], and xref:n1ql-rest-query:index.adoc#natural_context[natural_context] request-level parameters.
```

| **creds** +
__optional__
| Couchbase Capella credentials to authenticate the request.

Can be one of the following:

* A string in the username:password format. 

> * An object with the fields `user` and `pass`, similar to the xref:n1ql:n1ql-manage/query-settings.adoc#Credentials[creds] parameter.

Reviewer: ```suggestion
* An object with the fields `user` and `pass`, similar to the xref:n1ql:n1ql-manage/query-settings.adoc#Credentials[creds] request-level parameter.
```

> If specified, this value overrides the xref:n1ql-rest-query:index.adoc#Request[natural_cred] parameter.

Reviewer: ```suggestion
If specified, this value overrides the xref:n1ql-rest-query:index.adoc#natural_cred[natural_cred] request-level parameter.
```

Can be one of the following:

* A string matching the xref:n1ql-rest-query:index.adoc#Request[natural_context] specification. 
* An array of comma-separated strings.

> If specified, this value overrides the xref:n1ql-rest-query:index.adoc#Request[natural_context] parameter.

Reviewer 2: The xrefs to n1ql-rest-query for some reason seem to not work. If I click the link here I am redirected to https://preview.docs-test.couchbase.com/docs-devex-DOC-13536-using-ai-from-ui/server/current/n1ql/n1ql-language-reference/using-ai.html#n1ql-rest-query:index.adoc#Request 
______________________________________________________________

I was informed that this is something that will handle itself later.

Author: This is handled now and the redirects are working fine in the preview site.

| String or an array of strings 

| **orgId** +
__optional__

| Couchbase Capella organization ID for the request.

If specified, this value overrides the xref:n1ql-rest-query:index.adoc#Request[natural_orgid] parameter.

| String

| **execute** +
__optional__

| Indicates if the generated statement should be executed automatically. 

If `TRUE`, the Query Service executes the generated statement and returns the results. 
This applies only if the statement is a SELECT query.

> For other statement types, such as INSERT, UPDATE, DELETE, UPSERT, or CREATE FUNCTION, the statement is not executed, even if `execute` is `TRUE`.

Reviewer 2: nit: can we bold "not executed"?

Author: I understand why you'd want to highlight that portion, but as per Couchbase docs style guide, we can use bold only to format UI elements.

Can be one of the following:

> * A string matching the xref:n1ql-rest-query:index.adoc#Request[natural_context] specification. 

Reviewer: ```suggestion
* A string matching the xref:n1ql-rest-query:index.adoc#natural_context[natural_context] specification. 
```

* An array of comma-separated strings.

> If specified, this value overrides the xref:n1ql-rest-query:index.adoc#Request[natural_context] parameter.

Reviewer: ```suggestion
If specified, this value overrides the xref:n1ql-rest-query:index.adoc#natural_context[natural_context] request-level parameter.
```

| String or an array of strings 

| **orgId** +
__optional__

| Couchbase Capella organization ID for the request.

> If specified, this value overrides the xref:n1ql-rest-query:index.adoc#Request[natural_orgid] parameter.

Reviewer: ```suggestion
If specified, this value overrides the xref:n1ql-rest-query:index.adoc#natural_orgid[natural_orgid] request-level parameter.
```

To find your organization ID, log in to your Couchbase Capella account and check the URL in your web browser.
The organization ID is the `oid` parameter in the URL.

> For example, in the URL https://cloud.couchbase.com/databases?oid=5c670d3e-12a3-456b-7c89-123456789ab, the organization ID is `5c670d3e-12a3-456b-7c89-123456789ab`.

Reviewer: Let's make it not a link

```suggestion
For example, in the URL `+++https://cloud.couchbase.com/databases?oid=5c670d3e-12a3-456b-7c89-123456789ab+++`, the organization ID is `5c670d3e-12a3-456b-7c89-123456789ab`.
```

When set, these parameters apply to all subsequent USING AI statements in that session.

For example:

* In the cbq shell, you can set these parameters using the `\set` command. 
See <<example-1>>.
* In the Query Workbench, you can set these parameters in the xref:tools:query-workbench.adoc#query-preferences[Run-Time Preferences] window as named parameters, without the $ prefix. 

For more information about how to set request-level parameters, see xref:n1ql:n1ql-manage/query-settings.adoc#section_nnj_sjk_k1b[Configure Queries].

Once configured, you do not need to specify `creds`, `orgId`, or `keyspaces` in the WITH clause.

[[inline-parameters]]
=== Set Parameters Inline

You can set the `creds`, `orgId`, and `keyspaces` options directly in the `WITH` clause of the USING AI statement. 
If specified, these values override the `natural_cred`, `natural_orgid`, and `natural_context` parameters. 
See <<example-3>>.

== Result

A JSON object containing the generated statement, execution status, and metrics.

If the generated statement is a SELECT query and the `execute` option is `TRUE` (default), the output includes the query results as well.

> IMPORTANT: Currently, when you run a USING AI statement from the Query Workbench, it does not return the generated SQL++ statement. 

Reviewer: This is clearly not desired behavior. Check with PM if we should be highlighting this, especially if it's going to be fixed?

Author: Discussed with PMs and removed all lines and examples that highlight this issue.

Instead, it returns only the executed results of the generated statement. 

[[handling-passwords]]
== Handling Passwords Securely

Use `creds` and `natural_cred` parameters with caution to avoid exposing password information through history files or logs.
When working with the cbq shell, avoid passing passwords directly on the command line. 

> Instead, use the `\set` command to specify only the username, and then enter the password at the terminal prompt.

Reviewer 2: Can we show an example of \set -natural_cred with only the username:
```
cbq> \set -natural_cred gaurav.jayaraj@couchbase.com;
Enter password for "natural_cred":
cbq> \set;
 Query Parameters : 
 Parameter name : natural_cred
 Value : [gaurav.jayaraj@couchbase.com:***]

 Parameter name : profile
 Value : ["timings"]

 Named Parameters : 

 User Defined Session Parameters : 

 Predefined Session Parameters : 
 Parameter name : batch
 Value : ["off"]

 Parameter name : histfile
 Value : [".cbq_history"]

 Parameter name : pager
 Value : [false]

 Parameter name : quiet
 Value : [false]

 Parameter name : terse
 Value : [false]
```

Author: Done. Added a similar example.

Once configured, you do not need to specify `creds`, `orgId`, or `keyspaces` in the WITH clause.

[[inline-parameters]]
=== Set Parameters Inline

You can set the `creds`, `orgId`, and `keyspaces` options directly in the `WITH` clause of the USING AI statement. 
If specified, these values override the `natural_cred`, `natural_orgid`, and `natural_context` parameters. 
See <<example-3>>.

== Result

A JSON object containing the generated statement, execution status, and metrics.

If the generated statement is a SELECT query and the `execute` option is `TRUE` (default), the output includes the query results as well.

[[handling-passwords]]
== Handling Passwords Securely

Use `creds` and `natural_cred` parameters with caution to avoid exposing password information through history files or logs.
When working with the cbq shell, avoid passing passwords directly on the command line. 
Instead, use the `\set` command to specify only the username, and then enter the password at the terminal prompt.
This ensures that the password is not recorded in the shell history.

For example: 

> [source, sh]

Reviewer: I don't like specifying this as a source block because you get the Copy button, which implies you can just copy the example and paste it into your terminal. Which won't work, because this example is commands and responses: the whole conversation. So just format it as a listing block

```suggestion
```

----
cbq> \set -natural_cred username@example.com;
Enter password for "natural_cred": <enter password>

cbq> \set;
 Query Parameters : 
 Parameter name : natural_cred
 Value : [username@example.com:***]

 Parameter name : profile
 Value : ["timings"]

 ...
----

Similarly, if you're sending requests directly to a REST endpoint from a shell, be mindful of how you provide the password and whether it will be recorded in the shell history.
Consider using a method like the following to prompt for the password:

> [source,console]
> ----

Reviewer: Similarly let's just use a listing block here

```suggestion
----
```

echo -n "Enter your password: "
read -s p
echo
curl -s -d "natural_cred=<your-username>:${p}" ...
----

If you choose to reuse the password by setting an environment variable, note that it might be visible to other users on the system with sufficient privileges for process inspection (for example, through `/proc`).
If security is a concern, consider using an HTTPS connection. 

== Examples

In the following examples, replace `natural_cred`/`creds` and `natural_orgid`/`orgId` with your Couchbase Capella credentials and organization ID, respectively.

[[example-1]]
.USING AI with default request parameters
====
{example-note}

.Request

> [source, sh]

Reviewer: This might seem counterintuitive, but I actually use `sqlpp` as the file type for cbq shell commands. Because they actually are (mainly) SQL++ queries with interspersed shell commands. See the [cbq](https://docs.couchbase.com/server/current/n1ql/n1ql-intro/cbq.html) documentation for consistency.

```suggestion
[source, sqlpp]
```

====

[[example-2]]
.USING AI with keyspace specified within the statement and execute set to FALSE
====
{example-note}

.Request

> [source,sh]

Reviewer: ```suggestion
[source,sqlpp]
```

====

[[example-3]]
.USING AI with all parameters directly in the statement
====
{example-note}

.Request

> [source, sh]

Reviewer: ```suggestion
[source, sqlpp]
```

====

[[example-4]]
.USING AI with the Query REST API
====
.Request

> [source, console]

Reviewer: This one was right the first time

```suggestion
[source, sh]
```

{example-note}

.Request

> [source, sh]

Reviewer: ```suggestion
[source, sqlpp]
```

    }
}
----
If you examine the `hotel` keyspace, you'll see that this document was not inserted, even though `execute` was set to `TRUE`.
This is because the Query Service executes the generated statement only if it's a SELECT query.
====

[[example-6]]

> .USING AI with the flex index hint

Reviewer: Capitalization of terminology?

```suggestion
.USING AI with the Flex Index hint
```

====
{example-note}

.Request

> [source, sh]

Reviewer: ```suggestion
[source, sqlpp]
```

====
{example-note}

This example uses the same prompt as <<example-5>>, but specifies the `output` option instead of the `FLEXINDEX` keyword.

.Request

> [source, sh]

Reviewer: ```suggestion
[source, sqlpp]
```

{example-note}

.Request

> [source, sh]

Reviewer: ```suggestion
[source, sqlpp]
```

====

[[example-9]]
.Prefixing USING AI with EXPLAIN
====
{example-note}

.Request

> [source, sh]

Reviewer: ```suggestion
[source, sqlpp]
```
